### PR TITLE
Add type application in the type parser (#177)

### PR DESCRIPTION
**Pull request by wandernauta**
This parses every combination of a type variable followed by something else
as a type application.

Also adds the Functor type class and the fmap function to the catalog.

### CODE REVIEW

**Comment by comonoidial** [comment]
Also for '[]'  special case is needed somewhere in the type building process to ensure it ends up as an ListTypeCon

**Comment by wandernauta** [comment]
Is the arity of a tuple contstructor the number of arguments or the number of comma's? Those seem to be off by one.

**Comment by comonoidial** [comment]
Arity is the term refering to the number of arguments.
And the current TupleCon code is inconsistent with that.

**Comment by wandernauta** [comment]
The type builder now uses tuple and list constructors.

**Comment by comonoidial** [comment]
Roundtrips should be added to TypeBuilderTest for at least: () ,  (,)  , (,,) and [].

File: Code/src/test/java/nl/utwente/group10/haskell/typeparser/TypeBuilderTest.java [hunk 1]
```diff
@@ -29,16 +29,22 @@ private void roundtrip(String hs) {
     @Test public void testMaybeMaybe()  { this.roundtrip("Maybe (Maybe a)"); }
     @Test public void testMaybeSink()   { this.roundtrip("Maybe Int -> Maybe [a]"); }
     @Test public void testKitchenSink() { this.roundtrip("[a] -> [b] -> [(a, b)]"); }
+    @Test public void testPrefixUnit()  { this.roundtrip("()"); }
+    @Test public void testPrefixTuple() { this.roundtrip("(,) a"); }
+    @Test public void testPrefixTriple(){ this.roundtrip("(,,) a b"); }
+    @Test public void testPrefixList()  { this.roundtrip("[] a"); }
```
Comment (comonoidial) [inline Code/src/test/java/nl/utwente/group10/haskell/typeparser/TypeBuilderTest.java]: Why does roundtrip("[] a") succeed?
I would expect that "[] a" comes out as the special cased haskell type "[a]"

**Comment by comonoidial** [comment]
Looks all good to me :+1:
It's very nice that most of the standard libraries could used now without reimplementing type classes.

**Comment by wandernauta** [comment]
Merged, thanks for the hints @comonoidial.

**Pull request merged**